### ISSUE TITLE
hawtio3.x not able to connect remote jolokia

### ISSUE THREAD

**Issue by debasishdx08** [hawtio/hawtio#3401]
Hi,

I tried connecting jolokia application deployed on another server with the hawtio3.0.1.war and it loops infinitely.

The main reason for this is that while discovering local mbeans, the response returned is has the following format: `domains:{mbeans...}` however, when the response is fetched from external jolokia agent the response returned has the following format `{mbeans} `

Hence the `unwindListResponse` method fails to return any response for the remote response and the UI loops infinitely.

Is there any specific configuration required or is there any issue with remote jolokia connectivity for this version ?

**Comment by tadayosi** [comment]
Hi,

We need a bit more info for troubleshooting.

- Which web container are you using?
- What is the external JVM that you are trying to connect to?
- Which version of Jolokia are you using for the external JVM?
- Can you provide us with more concrete response of `domains:{mbeans...}`?

**Comment by grgrzybek** [comment]
I agree with @tadayosi, we need more information...

> I tried connecting jolokia application deployed on another server with the hawtio3.0.1.war and it loops infinitely.

Please be more specific - I read it several times, but I still don't know whether the WAR is the application you're connecting _from_ or connecting _to_. And this _another server_ is what kind of server? What do you mean by "loop" - a Java StackOverflowException? Or blank page (which may seem like waiting for something)...

Can you attach the JSON fragment with `domains:{mbeans...}`?

**Comment by debasishdx08** [comment]
Both source and target are JBoss wildfly with the following deployment - source has hawtio.war along with fileloading.app.ear deployed, target has jolokia.war along with fileloading.app.ear deployed.
Source server deployment screenshot:
![image](https://github.com/hawtio/hawtio/assets/58969134/3f6b6f7b-03c0-4093-a913-5b4415f69160)

Following is the `mbean `content that comes withing `domains: {mbeans...}` when hawito is doing `list()` on source server itself.
![image](https://github.com/hawtio/hawtio/assets/58969134/66e92812-8d6e-4e95-863e-8c2fb04c1fb0)

Chrome debugger screenshot for response of source JVM:
![image](https://github.com/hawtio/hawtio/assets/58969134/edd478e1-1f8b-4a08-880a-697fb26a4503)

Target server deployment screenshot:
![image](https://github.com/hawtio/hawtio/assets/58969134/1e48c399-cf09-45e6-93e2-a7184e41cb18)

The same value is returned from the target server when jolokia list is called on the remote server but without wrapped within `domains` and the screen keeps loading, screenshot below:
![image](https://github.com/hawtio/hawtio/assets/58969134/1d871078-a59f-4626-84f6-57888b1b400f)

Chrome debugger screenshot for response of target JVM:
![image](https://github.com/hawtio/hawtio/assets/58969134/a9382e59-7db9-4a4d-885a-44c1ca3cab1f)

The jolokia.war(1.7.2) deployed on the target server has jokolia-core-1.7.2.jar and the hawtio3.0.1 war deployed on the source has the same jolokia-core-1.7.2.jar.

Following error is thrown while trying to extract mbean information withing the `unwindListResponse` method because it expects the response needs to be wrapped with `domains: {mbeans}`
![image](https://github.com/hawtio/hawtio/assets/58969134/1d3bb676-bab3-4c0b-8043-c42be5375e21)

Please let me know if any further details are needed. Thanks in advance !

**Comment by grgrzybek** [comment]
> because it expects the response needs to be wrapped with `domains: {mbeans}`

Not necessarily - this is one of potential responses. Such value (with `domains` and `cache`) are possible only for Karaf deployments with optimization for Camel and ActiveMQ beans (and you don't seem to have any).

You can export the network tab content to a HAR file (from Firefox or Chrome) - is it possible to get such dump?

**Comment by debasishdx08** [comment]
Dump for local(source_dump.har) and remote(target_dump.har) Mbean listing are attached here:
[dumps.zip](https://github.com/hawtio/hawtio/files/15073163/dumps.zip)

**Comment by grgrzybek** [comment]
Which Wildfly version are you using?
Are you simply connecting from one hawtio to another using /connect/remote ("Connect" tab)? What did you provide in URL of the remote agent:
![image](https://github.com/hawtio/hawtio/assets/306565/33196180-f7ad-44a7-be6f-cabebd1b3c28)

**Comment by grgrzybek** [comment]
With hawtio-3.0.1.war dropped into Wildfly 26.1.3's standalone/deployments, everything works. Also connection to _self_ (http://localhost:8080/hawtio-war-3.0.1/jolokia) works fine.

I thought you have two WildFlies running, each with hawtio-3.0.1 (that was my first impression). But you're connecting to standalone Jolokia agent - how did you configure this agent (startup command) and what does this command return?:

```
curl -i -XPOST -H'Content-Type: application/json' -H'Accept: application/json' -d '{"type":"list"}' http://host:8750/jolokia/list?max_depth=7
```

**Comment by debasishdx08** [comment]
Yes, you are correct. If I deploy the hawtio.war in target as well, then it works fine. It does not work when the target has just the jolokia.war deployed instead of the hawtio.war. The target WildFly standalone/deployment looks like below:

![image](https://github.com/hawtio/hawtio/assets/58969134/295921e4-fbe2-4238-a892-7cb3a05d6d01)

The response of the curl command is shown below:
![image](https://github.com/hawtio/hawtio/assets/58969134/eaf4960d-2438-4320-bc44-ed3925d86ac1)

The jolokia-unsecured.war (renamed as jolokia.war) is deployed on the target WildFly for basic testing and evaluation.

**Comment by grgrzybek** [comment]
Because I'm not that fluent with wildfly configuration - where did you configure the port 8750? `socket-binding-group` in `standalone/configuration/standalone.xml`? Any other changes there?

**Comment by debasishdx08** [comment]
Yes, the changes are needed in the `socket-binding-group` section. The default value are 8080 for http and 8443 for https. The section `port-offset="${jboss.socket.binding.port-offset:0}"` defines the offset value needs to be added where we have passed 670 as the offset hence the http is listening on 8080 + 670 = 8750.

With the default settings, you can try with 8080/ 8443. Depending on if you have one or both enabled, you should be able to get the response.

**Comment by debasishdx08** [comment]
Did you manage to replicate the issue ?

**Comment by grgrzybek** [comment]
Not yet... Everything worked fine. I had some theory about number of MBeans, but couldn't verify it. Is `fileloading.app.ear` something I could install? Or is it your private application?

**Comment by debasishdx08** [comment]
It's private application. You want me to try something out ?

**Comment by grgrzybek** [comment]
no - for now I'll try to simulate. thanks!

**Comment by grgrzybek** [comment]
As expected - works with jolokia.war running in the other Wildfly instance...

I'll try to deploy an app with simulated MBeans registered (from your HAR).

**Comment by debasishdx08** [comment]
We are running on `open-jdk 17`, can that potentially cause any issues ? Also the jolokia war in you case is listening on `/jolokia` context not `/hawtio/jolokia` correct ?

**Comment by grgrzybek** [comment]
JDK shouldn't matter. Also standalone jolokia can run without problems on `/jolokia` when accessed from another Hawtio.

Please remind me - you see a problem in _main_ Hawtio or in the Hawtio opened with "Connect" button?

**Comment by grgrzybek** [comment]
For now, I've extracted 2764 MBean names which I'm going to create in both Wildflies. Perhaps some of these MBean names cause problems?

**Comment by grgrzybek** [comment]
I've registered all the MBeans from your HAR files (`source-dump.har`), but without those in `jboss.as` domain (you can't do it). I got your MBeans (these are fake objects without any attributes/methods/notifications), but it still works...
![image](https://github.com/hawtio/hawtio/assets/306565/a0f3be77-f873-4037-b468-46e88dbedf26)

**Comment by debasishdx08** [comment]
I tried without deploying the `fileloading.ear` I am able to connect, but when I deploy the `fileloading.ear` then it fails to connect. `isJmxDomains(response)` returns true without the fileloading and returns false with my ear. Not sure what to check now

```
const isJmxDomain = (value: unknown): value is OptimisedJmxDomain =>
      is(value, record(string(), define('MBeanInfo', isMBeanInfo)))
    const isJmxDomains = (value: unknown): value is OptimisedJmxDomains =>
      is(value, record(string(), define('JmxDomain', isJmxDomain)))

    if (isOptimisedListResponse(response)) {
      // Post process cached MBean info
      const { cache, domains } = response
      Object.entries(domains).forEach(([_, domain]) => {
        Object.entries(domain).forEach(([mbeanName, mbeanOrCache]) => {
          if (isString(mbeanOrCache)) {
            domain[mbeanName] = cache[mbeanOrCache] as OptimisedMBeanInfo
          }
        })
      })
      return domains as OptimisedJmxDomains
    }

    if (isJmxDomains(response)) {
      return response
    }
```

**Comment by grgrzybek** [comment]
Oh, good hint about domains true/false!

And the good news is that I've reproduced it! Taking your HAR, I've extracted Jolokia's `type: list` response and I've tweaked Jolokia 1.7.2 to simply return this JSON instead of searching current MBeanServer and I got it!:
![image](https://github.com/hawtio/hawtio/assets/306565/68489115-8cf2-41cd-8e21-e8bd6459c978)

**Comment by debasishdx08** [comment]
> For the record - it also fails on latest Hawtio 4.

It was working in hawtio 2.x with jolokia 1.7.2

**Comment by debasishdx08** [comment]
Thanks a lot !

**Comment by grgrzybek** [comment]
Now it works!
